### PR TITLE
Improved FileFinder Search Speed

### LINKED ISSUE #30408: ALC File Search Speed Improvements
Currently if a user wanted to load say 200 runs, the interface would search for all 200 runs. Instead of doing this, it has been suggested we only need to search for the start and end files of a range and any comma separated files, e.g. Entering `62260-5,62267,62269,62271-75`

Previous search would be for all **13** files, instead we want to only search for **6** files; 62260, 62265, 62267, 62269, 62271, 62275. 

This should hopefully have big performance improvements particularly when entering large ranges which can reduce searches of tens or hundreds of files to just 2. This will only work on the assumption that the files in between the start and end runs of ranges all have the same directory such that we can construct the rest of the paths. 

I have put together some of the potential changes that could/should be made:
1. Instead of using the standard filefinderwidget, would want to potentially create an alc specific one 
2. Should be hesitant at creating a whole new widget as the current filefinder handles multithreading and archive searching, so maybe just inheriting is the way forward
3. You would then want to override the function `findFilesGetSearchText` to return the reduced number of files (see example above). Make sure to check what format is needed for the search. A problem with this is knowing which files are part of a range and which are not, as we would need to ensure the start and end files have been found in the same directory. separate ranges or files can be loaded from different directories to allow for comparing different cycles (PlotAsymmetryByLogValue would need to be changed as there is a check all files are in same directory currently)
4. There will be a problem when calling `getFilenames` on the filefinderwidget, as the filesfound will no longer be all the files wanting to be loaded, so from the files found you would have to construct the rest of the files to be loaded
5. Once you have constructed all the files the rest should be the same, load times should not change

By creating a new widget for finding files it may be easier to solve #30067

I tried loading MUSR62260-62460 and to search for the files took ≈50 seconds, so any changes made should aim to be significantly faster than this

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
When searching for multiple runs, added a check to see if the file exists using the current run number and the previous path and extension. This provides a great speed benefit particularly if the runs are stored on the archive in one directory, which is the most common use case, by limiting calls made to the web api.

**To test:**
I have been using this script for testing. But you should see overall improvements when loading runs in interfaces as well (default instrument EMU)
```
# Worst case every file in different directory 
archive_search_10_dirs = "70993, 72589, 74374, 76366, 78873, 83480, 86490, 88127, 91955, 95162" 

# Baseline
ten_archive_files = "70993-71003" 

# Some cases interfaces would like to load in hundreds of runs and were limited by long search times
one_hundred_archive_files = "70993-71093" 
two_hundred_archive_files = "70993-71193" 

# I have also been testing local files but you will need to set this up yourself

n_times = 3 # how many times to run the search
average = 0

for i in range(1, n_times+1):
    start = time.time()
    runs = FileFinder.findRuns(hintstr=one_hundred_archive_files) # Add in what you are searching for here as the parameter hintstr
    end = time.time()
    print("Elapsed time {}: {} seconds".format(i, end-start))
    average += end-start

print("Average time: {}".format(average/n_times))
```

Fixes #30408 

These are the results from my own testing with and without the changes from this PR:
```
Without Changes:
10 files in 10 different archive directories* - 
	Elapsed time 1 2.343200445175171 seconds
	Elapsed time 2 2.2812581062316895 seconds
	Elapsed time 3 2.296821117401123 seconds
	Average time 2.3070932229359946
10 files in 10 different local directories - 
	Elapsed time 1 0.12475180625915527 seconds
	Elapsed time 2 0.01560068130493164 seconds
	Elapsed time 3 0.03126859664916992 seconds
	Average time 0.05720702807108561
10 files in 1 archive directory -
	Elapsed time 1 2.7340285778045654 seconds
	Elapsed time 2 2.5316970348358154 seconds
	Elapsed time 3 2.7501251697540283 seconds
	Average time 2.6719502607981362
10 files in 1 local directory - 
	Elapsed time 1 0.0156552791595459 seconds
	Elapsed time 2 0.0 seconds
	Elapsed time 3 0.015536069869995117 seconds
	Average time 0.010397116343180338
100 files in 1 archive directory - 
	Elapsed time 1 24.585033178329468 seconds
	Elapsed time 2 25.843957901000977 seconds
	Elapsed time 3 26.253194093704224 seconds
	Average time 25.560728391011555
200 files in 1 archive directory - 
	Elapsed time 1 49.61705946922302 seconds
	Elapsed time 2 49.03502941131592 seconds
	Elapsed time 3 48.783374547958374 seconds
	Average time 49.14515447616577

With Changes:
10 files in 10 different archive directories* - 
	Elapsed time 1 2.342933416366577 seconds
	Elapsed time 2 2.2974653244018555 seconds
	Elapsed time 3 2.2344136238098145 seconds
	Average time 2.2916041215260825
10 files in 10 different local directories - 
	Elapsed time 1 0.062261104583740234 seconds
	Elapsed time 2 0.031131982803344727 seconds
	Elapsed time 3 0.04710721969604492 seconds
	Average time 0.046833435694376625
10 files in 1 archive directory - 
	Elapsed time 1 0.3748657703399658 seconds
	Elapsed time 2 0.2499403953552246 seconds
	Elapsed time 3 0.20333147048950195 seconds
	Average time 0.2760458787282308
10 files in 1 local directory - 
	Elapsed time 1 0.10909771919250488 seconds
	Elapsed time 2 0.0 seconds
	Elapsed time 3 0.0 seconds
	Average time 0.036365906397501625
100 files in 1 archive directory - 
	Elapsed time 1 0.5618562698364258 seconds
	Elapsed time 2 0.266071081161499 seconds
	Elapsed time 3 0.2651832103729248 seconds
	Average time 0.3643701871236165
200 files in 1 archive directory - 
	Elapsed time 1 0.7649815082550049 seconds
	Elapsed time 2 0.2968416213989258 seconds
	Elapsed time 3 0.31253862380981445 seconds
	Average time 0.45812058448791504
* Using (EMU 70993, 72589, 74374, 76366, 78873, 83480, 86490, 88127, 91955, 95162)
```
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/FileFinder.cpp [hunk 1]
```diff
@@ -605,12 +605,32 @@ FileFinderImpl::findRuns(const std::string &hintstr,
       if (runEndNumber < runNumber) {
         throw std::invalid_argument("Malformed range of runs: " + *h);
       }
+      std::string previous_path{""};
```
Comment (martyngigg) [inline Framework/API/src/FileFinder.cpp:608]: For strings the default constructor will give an empty string so we can collapse this to:
```c++
std::string previousPath, previousExt;
```
Note that I've switched to camelCase to be consistent with the coding standards.

File: Framework/API/src/FileFinder.cpp [hunk 2]
```diff
@@ -605,12 +605,32 @@ FileFinderImpl::findRuns(const std::string &hintstr,
...
+              continue;
+            }
+          } catch (...) {
+          }
+        }
+
         std::string path = findRun(p1.first + run, exts, useExtsOnly);
         if (!path.empty()) {
+          // Cache successfully found path and extension
+          previous_ext = "." + Poco::Path(path).getExtension();
```
Comment (martyngigg) [inline Framework/API/src/FileFinder.cpp:632]: I think the Poco::Path constructor is a little expensive so it's probably worth a separate variable for that first.

File: docs/source/release/v6.1.0/framework.rst [hunk 3]
```diff
@@ -24,6 +24,7 @@ Data Objects
 ------------
 
 - :ref:`Run <mantid.api.Run>` has been modified to allow multiple goniometers to be stored.
+- :ref:`FileFinder <mantid.api.FileFinderImpl>` has been modified to improve search times in some cases.
```
Comment (martyngigg) [inline docs/source/release/v6.1.0/framework.rst:27]: Can we be a little more explicit in the *some* cases? Perhaps just say in the case of searching for multiple runs for the same instrument?

**Review by martyngigg — CHANGES_REQUESTED** [review]
I like this improvement as it uses a fact that is true much of the time for facility directories where runs are typically next to each other. Offline you showed some timings you had done. Could you include them in the PR description for future reference?

Just a couple of inline code comments to address too.

**Comment by martyngigg** [comment]
@gvardany Might this help with the issues in #29881 or is that more generally an issue for multiple files and not just runs?

**Comment by gvardany** [comment]
This seems indeed very similar. I will try with my script on this branch to see if it helps. It is just that my concern was finding via MultipleFileProperty, but eventually this must rely on the same FileFinder anyhow.

**Comment by martyngigg** [comment]
> This seems indeed very similar. I will try with my script on this branch to see if it helps. It is just that my concern was finding via MultipleFileProperty, but eventually this must rely on the same FileFinder anyhow.

Yes I hope so. If not we should probably try and have these things converge.

File: Framework/API/src/FileFinder.cpp [hunk 4]
```diff
@@ -605,12 +605,31 @@ FileFinderImpl::findRuns(const std::string &hintstr,
...
+        // path/extension
+        if (!previousPath.empty() && !previousExt.empty()) {
+          try {
+            const Poco::File file(previousPath + p1.first + run + previousExt);
+            if (file.exists()) {
+              res.emplace_back(previousPath + p1.first + run + previousExt);
+              continue;
+            }
+          } catch (...) {
+          }
```
Comment (martyngigg) [inline Framework/API/src/FileFinder.cpp:624]: If we have caught an error here perhaps we should clear the previous path and extensions as they likely causes issues if the next one is okay it'll cache again anyway?

File: Framework/API/src/FileFinder.cpp [hunk 5]
```diff
@@ -605,12 +605,31 @@ FileFinderImpl::findRuns(const std::string &hintstr,
...
       for (int irun = runNumber; irun <= runEndNumber; ++irun) {
         run = std::to_string(irun);
         while (run.size() < nZero)
           run.insert(0, "0");
+
+        // Quick check if file can be created from previous successfully found
+        // path/extension
+        if (!previousPath.empty() && !previousExt.empty()) {
+          try {
+            const Poco::File file(previousPath + p1.first + run + previousExt);
```
Comment (martyngigg) [inline Framework/API/src/FileFinder.cpp:618]: Can we not duplicate `previousPath + p1.first + run + previousExt` below? From a readability point of you I need to double take to check that they are the same.

I think `Poco::File` has a string method?

**Review by martyngigg — APPROVED** [review]
I think this is an excellent improvement and is ready to go now.

**Comment by gvardany** [comment]
Unfortunately this didn't improve my observations on our archive (https://github.com/mantidproject/mantid/issues/29881).
Here is what I got before and after the merge of this PR.
3rd and 4th entries are when setting the same value to the same property, those are faster due to caching of its previous value in multiple file property.
Still I have hard time interpreting these timings.
I'm puzzled that even with caching the resolved paths it takes around 60sec.
This means that it is probably the time spent just checking if the files exists w/o any finding.
I'm also confused why for the first time setting with absolute paths takes longer than with numors.
I had a draft attempt https://github.com/mantidproject/mantid/tree/multiplefiles_samedirectory based on the assumption that in one go all the numors must be in the same directory, but that didn't seem to change the picture either.

```
Master (after merge)
--------------------
Absolute paths took: 126.40826392173767
Numors took: 94.69144296646118
Absolute paths took: 62.48769807815552
Numors took: 61.87765407562256

Master (before merge)
---------------------
Absolute paths took: 122.73527884483337
Numors took: 101.9387719631195
Absolute paths took: 61.36741924285889
Numors took: 61.29860305786133
```

**Pull request merged**